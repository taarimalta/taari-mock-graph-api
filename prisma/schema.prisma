generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

model User {
  id                         Int                @id @default(autoincrement())
  username                   String             @unique
  email                      String             @unique
  firstName                  String?
  lastName                   String?
  createdAt                  DateTime           @default(now())
  modifiedAt                 DateTime           @default(now())
  createdBy                  Int?
  modifiedBy                 Int?
  animalsModified            Animal[]           @relation("AnimalModifiedBy")
  animalsCreated             Animal[]           @relation("AnimalCreatedBy")
  countriesModified          Country[]          @relation("CountryModifiedBy")
  countriesCreated           Country[]          @relation("CountryCreatedBy")
  domainsModified            Domain[]           @relation("DomainModifiedBy")
  domainsCreated             Domain[]           @relation("DomainCreatedBy")
  modifier                   User?              @relation("UserModifiedBy", fields: [modifiedBy], references: [id])
  modifiedUsers              User[]             @relation("UserModifiedBy")
  creator                    User?              @relation("UserCreatedBy", fields: [createdBy], references: [id])
  createdUsers               User[]             @relation("UserCreatedBy")
  modifiedUserDomainAccesses UserDomainAccess[] @relation("UserDomainAccessModifiedBy")
  createdUserDomainAccesses  UserDomainAccess[] @relation("UserDomainAccessCreatedBy")
  domainAccess               UserDomainAccess[]
}

model Country {
  id         Int       @id @default(autoincrement())
  name       String
  capital    String?
  population Int?
  area       Float?
  currency   String?
  continent  String
  createdAt  DateTime  @default(now())
  modifiedAt DateTime?
  createdBy  Int?
  modifiedBy Int?
  modifier   User?     @relation("CountryModifiedBy", fields: [modifiedBy], references: [id])
  creator    User?     @relation("CountryCreatedBy", fields: [createdBy], references: [id])
}

model Animal {
  id                  Int       @id @default(autoincrement())
  name                String
  species             String?
  habitat             String?
  diet                String?
  conservation_status String?
  category            String
  createdAt           DateTime  @default(now())
  modifiedAt          DateTime?
  createdBy           Int?
  modifiedBy          Int?
  modifier            User?     @relation("AnimalModifiedBy", fields: [modifiedBy], references: [id])
  creator             User?     @relation("AnimalCreatedBy", fields: [createdBy], references: [id])
}

model Domain {
  id         Int                @id @default(autoincrement())
  name       String
  parentId   Int?
  createdAt  DateTime           @default(now())
  modifiedAt DateTime?
  createdBy  Int?
  modifiedBy Int?
  modifier   User?              @relation("DomainModifiedBy", fields: [modifiedBy], references: [id])
  creator    User?              @relation("DomainCreatedBy", fields: [createdBy], references: [id])
  parent     Domain?            @relation("DomainParent", fields: [parentId], references: [id])
  children   Domain[]           @relation("DomainParent")
  userAccess UserDomainAccess[]
}

model UserDomainAccess {
  id         Int      @id @default(autoincrement())
  userId     Int
  domainId   Int
  createdAt  DateTime @default(now())
  modifiedAt DateTime @default(now())
  createdBy  Int?
  modifiedBy Int?
  modifier   User?    @relation("UserDomainAccessModifiedBy", fields: [modifiedBy], references: [id])
  creator    User?    @relation("UserDomainAccessCreatedBy", fields: [createdBy], references: [id])
  domain     Domain   @relation(fields: [domainId], references: [id], onDelete: Cascade)
  user       User     @relation(fields: [userId], references: [id], onDelete: Cascade)

  @@unique([userId, domainId])
}
