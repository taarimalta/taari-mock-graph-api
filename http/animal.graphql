# Sample GraphQL queries and mutations for Animal

# Query all animals
query GetAnimals {
  animals {
    id
    name
    classification
  }
}

# Query a single animal by ID
query GetAnimal {
  animal(id: 1) {
    id
    name
    classification
  }
}

# Create a new animal
mutation CreateAnimal {
  createAnimal(name: "Tiger", classification: Mammal) {
    id
    name
    classification
  }
}

# Update an animal
mutation UpdateAnimal {
  updateAnimal(id: 1, name: "Tiger Updated", classification: Mammal) {
    id
    name
    classification
  }
}

# Delete an animal
mutation DeleteAnimal {
  deleteAnimal(id: 1) {
    id
    name
    classification
  }
}
