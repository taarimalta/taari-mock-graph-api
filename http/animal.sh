#!/bin/bash
# Test Animal GraphQL API with seeded data

API_URL="http://localhost:4000/graphql"

# Check if server is running by sending a trivial POST query
HEALTH_JSON=$(curl -s -X POST -H "Content-Type: application/json" --data '{"query":"{ __typename }"}' "$API_URL")
if ! echo "$HEALTH_JSON" | jq -e .data > /dev/null; then
  echo "‚ùå API server is not running or not responding correctly at $API_URL"
  exit 1
fi

# Query all animals and extract IDs
ANIMALS_JSON=$(curl -s -X POST -H "Content-Type: application/json" --data '{
  "query": "{ animals { id name category species habitat diet conservation_status } }"
}' $API_URL)
echo "All animals:"
echo "$ANIMALS_JSON" | jq .

# Extract IDs for each category
declare -A CATEGORY_IDS
for CATEGORY in mammals birds reptiles amphibians fish insects; do
  ID=$(echo "$ANIMALS_JSON" | jq -r ".data.animals[] | select(.category==\"$CATEGORY\") | .id" | head -n1)
  CATEGORY_IDS[$CATEGORY]=$ID
done

# Query by each category's animal ID
echo "\nQuerying one animal from each category:"
for CATEGORY in "${!CATEGORY_IDS[@]}"; do
  ID=${CATEGORY_IDS[$CATEGORY]}
  if [ -n "$ID" ]; then
    echo "- $CATEGORY (id=$ID):"
    curl -s -X POST -H "Content-Type: application/json" --data '{
      "query": "{ animal(id: '$ID') { id name category species habitat diet conservation_status } }"
    }' $API_URL | jq .
  fi
done

# Update the first animal (mammals)
MAMMAL_ID=${CATEGORY_IDS[mammals]}
if [ -n "$MAMMAL_ID" ]; then
  echo "\nUpdating mammal (id=$MAMMAL_ID)..."
  curl -s -X POST -H "Content-Type: application/json" --data '{
    "query": "mutation { updateAnimal(id: '$MAMMAL_ID', name: \"Elephant Updated\") { id name category } }"
  }' $API_URL | jq .
fi

# Delete the first animal (mammals)
if [ -n "$MAMMAL_ID" ]; then
  echo "\nDeleting mammal (id=$MAMMAL_ID)..."
  curl -s -X POST -H "Content-Type: application/json" --data '{
    "query": "mutation { deleteAnimal(id: '$MAMMAL_ID') { id name } }"
  }' $API_URL | jq .
fi
