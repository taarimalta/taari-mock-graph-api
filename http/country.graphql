# Sample GraphQL queries and mutations for Country

# Query all countries
query GetCountries {
  countries {
    id
    name
    continent
  }
}

# Query a single country by ID
query GetCountry {
  country(id: 1) {
    id
    name
    continent
  }
}

# Create a new country
mutation CreateCountry {
  createCountry(name: "Nepal", continent: Asia) {
    id
    name
    continent
  }
}

# Update a country
mutation UpdateCountry {
  updateCountry(id: 1, name: "Nepal Updated", continent: Asia) {
    id
    name
    continent
  }
}

# Delete a country
mutation DeleteCountry {
  deleteCountry(id: 1) {
    id
    name
    continent
  }
}
