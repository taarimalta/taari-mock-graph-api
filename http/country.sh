#!/bin/bash
# Test Country GraphQL API with seeded data

API_URL="http://localhost:4000/graphql"


# Check if server is running by sending a trivial POST query
HEALTH_JSON=$(curl -s -X POST -H "Content-Type: application/json" --data '{"query":"{ __typename }"}' "$API_URL")
if ! echo "$HEALTH_JSON" | jq -e .data > /dev/null; then
  echo "‚ùå API server is not running or not responding correctly at $API_URL"
  exit 1
fi

# Query all countries and extract IDs
COUNTRIES_JSON=$(curl -s -X POST -H "Content-Type: application/json" --data '{
  "query": "{ countriesPaginated(args: { first: 1000 }) { data { id name continent capital population area currency } } }"
}' $API_URL)
echo "All countries:"
echo "$COUNTRIES_JSON" | jq .

# Extract IDs for each continent
declare -A CONTINENT_IDS
for CONTINENT in africa asia europe northamerica southamerica oceania; do
  ID=$(echo "$COUNTRIES_JSON" | jq -r ".data.countries[] | select(.continent==\"$CONTINENT\") | .id" | head -n1)
  CONTINENT_IDS[$CONTINENT]=$ID
done

# Query by each continent's country ID
echo "\nQuerying one country from each continent:"
for CONTINENT in "${!CONTINENT_IDS[@]}"; do
  ID=${CONTINENT_IDS[$CONTINENT]}
  if [ -n "$ID" ]; then
    echo "- $CONTINENT (id=$ID):"
    curl -s -X POST -H "Content-Type: application/json" --data '{
      "query": "{ country(id: '$ID') { id name continent capital population area currency } }"
    }' $API_URL | jq .
  fi
done

# Update the first country (africa)
AFRICA_ID=${CONTINENT_IDS[africa]}
if [ -n "$AFRICA_ID" ]; then
  echo "\nUpdating African country (id=$AFRICA_ID)..."
  curl -s -X POST -H "Content-Type: application/json" --data '{
    "query": "mutation { updateCountry(id: '$AFRICA_ID', name: \"Nigeria Updated\") { id name continent } }"
  }' $API_URL | jq .
fi

# Delete the first country (africa)
if [ -n "$AFRICA_ID" ]; then
  echo "\nDeleting African country (id=$AFRICA_ID)..."
  curl -s -X POST -H "Content-Type: application/json" --data '{
    "query": "mutation { deleteCountry(id: '$AFRICA_ID') { id name } }"
  }' $API_URL | jq .
fi
