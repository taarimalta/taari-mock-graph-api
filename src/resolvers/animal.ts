import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

export const animalResolvers = {
  Query: {
    animals: (_: any, args: { search?: string, filter?: any }) => {
      // Deprecated: Use animalsPaginated
      const { buildAnimalWhere } = require('../utils/filtering');
      const where = buildAnimalWhere(args.filter, args.search);
      return prisma.animal.findMany({ where });
    },
    animalsPaginated: async (_: any, args: {
      search?: string,
      filter?: any,
      orderBy?: any,
      args?: any
    }) => {
      const { buildAnimalWhere } = require('../utils/filtering');
      const { mapAnimalOrderField, buildOrderBy } = require('../utils/sorting');
      const { paginate } = require('../utils/pagination');

      const where = buildAnimalWhere(args.filter, args.search);
      const orderField = args.orderBy?.field || 'NAME';
      const direction = args.orderBy?.direction || 'ASC';
      const orderBy = buildOrderBy(mapAnimalOrderField(orderField), direction);
      const pageArgs = args.args || { first: 20 };

      const result = await paginate({
        model: prisma.animal,
        where,
        orderBy,
        first: pageArgs.first,
        after: pageArgs.after,
        last: pageArgs.last,
        before: pageArgs.before,
      });
      return {
        data: Array.isArray(result.items) ? result.items : [],
        pagination: result.pagination,
      };
    },
    animal: (_: any, args: { id: number }) =>
      prisma.animal.findUnique({ where: { id: Number(args.id) } }),
  },
  Mutation: {
    createAnimal: (
      _: any,
      args: {
        name: string;
        species?: string;
        habitat?: string;
        diet?: string;
        conservation_status?: string;
        category: string;
      }
    ) =>
      prisma.animal.create({
        data: {
          name: args.name,
          species: args.species,
          habitat: args.habitat,
          diet: args.diet,
          conservation_status: args.conservation_status,
          category: args.category,
        },
      }),
    updateAnimal: (
      _: any,
      args: {
        id: number;
        name?: string;
        species?: string;
        habitat?: string;
        diet?: string;
        conservation_status?: string;
        category?: string;
      }
    ) =>
      prisma.animal.update({
        where: { id: Number(args.id) },
        data: {
          ...(args.name !== undefined ? { name: args.name } : {}),
          ...(args.species !== undefined ? { species: args.species } : {}),
          ...(args.habitat !== undefined ? { habitat: args.habitat } : {}),
          ...(args.diet !== undefined ? { diet: args.diet } : {}),
          ...(args.conservation_status !== undefined ? { conservation_status: args.conservation_status } : {}),
          ...(args.category !== undefined ? { category: args.category } : {}),
        },
      }),
    deleteAnimal: (_: any, args: { id: number }) =>
      prisma.animal.delete({ where: { id: Number(args.id) } }),
  },
};
