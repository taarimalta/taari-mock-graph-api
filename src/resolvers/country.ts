import { PrismaClient } from '@prisma/client';
const prisma = new PrismaClient();

export const countryResolvers = {
  Query: {
    countries: (_: any, args: { search?: string, filter?: any }) => {
      // Deprecated: Use countriesPaginated
      const { buildCountryWhere } = require('../utils/filtering');
      const where = buildCountryWhere(args.filter, args.search);
      return prisma.country.findMany({ where });
    },
    countriesPaginated: async (_: any, args: {
      search?: string,
      filter?: any,
      orderBy?: any,
      args?: any
    }) => {
      try {
        const { buildCountryWhere } = require('../utils/filtering');
        const { mapCountryOrderField, buildOrderBy } = require('../utils/sorting');
        const { paginate } = require('../utils/pagination');

        const where = buildCountryWhere(args.filter, args.search);
        const orderField = args.orderBy?.field || 'NAME';
        const direction = args.orderBy?.direction || 'ASC';
        const orderBy = buildOrderBy(mapCountryOrderField(orderField), direction);
        const pageArgs = args.args || { first: 20 };

        const result = await paginate({
          model: prisma.country,
          where,
          orderBy,
          first: pageArgs.first,
          after: pageArgs.after,
          last: pageArgs.last,
          before: pageArgs.before,
        });
        
        return {
          data: Array.isArray(result.items) ? result.items : [],
          pagination: result.pagination,
        };
      } catch (err) {
        console.error('countriesPaginated error:', err);
        return { data: [], pagination: { hasNext: false, hasPrevious: false, startCursor: null, endCursor: null, totalCount: 0 } };
      }
    },
    country: (_: any, args: { id: number }) =>
      prisma.country.findUnique({ where: { id: Number(args.id) } }),
  },
  Mutation: {
    createCountry: (
      _: any,
      args: {
        name: string;
        capital?: string;
        population?: number;
        area?: number;
        currency?: string;
        continent: string;
      }
    ) =>
      prisma.country.create({
        data: {
          name: args.name,
          capital: args.capital,
          population: args.population,
          area: args.area,
          currency: args.currency,
          continent: args.continent,
        },
      }),
    updateCountry: (
      _: any,
      args: {
        id: number;
        name?: string;
        capital?: string;
        population?: number;
        area?: number;
        currency?: string;
        continent?: string;
      }
    ) =>
      prisma.country.update({
        where: { id: Number(args.id) },
        data: {
          ...(args.name !== undefined ? { name: args.name } : {}),
          ...(args.capital !== undefined ? { capital: args.capital } : {}),
          ...(args.population !== undefined ? { population: args.population } : {}),
          ...(args.area !== undefined ? { area: args.area } : {}),
          ...(args.currency !== undefined ? { currency: args.currency } : {}),
          ...(args.continent !== undefined ? { continent: args.continent } : {}),
        },
      }),
    deleteCountry: (_: any, args: { id: number }) =>
      prisma.country.delete({ where: { id: Number(args.id) } }),
  },
};
