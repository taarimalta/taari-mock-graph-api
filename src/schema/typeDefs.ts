import { gql } from 'graphql-tag';

export const typeDefs = gql`
  """
  Opaque cursor token (base64-encoded payload from the last item on a page).
  Clients must treat this as an unreadable token.
  """
  scalar Cursor

  """
  All continents supported by the API for classification of countries.
  """
  enum Continent {
    africa
    asia
    europe
    northamerica
    southamerica
    oceania
  }

  """
  Categories for classifying animals.
  """
  enum AnimalCategory {
    mammals
    birds
    reptiles
    amphibians
    fish
    insects
  }

  """
  Country information including identifiers, demographics, and geography.
  """
  type Country {
    id: ID!
    name: String!
    capital: String
    population: Int
    area: Float
    currency: String
    continent: Continent!
  }

  """
  Animal information including biological, ecological, and conservation details.
  """
  type Animal {
    id: ID!
    name: String!
    species: String
    habitat: String
    diet: String
    conservation_status: String
    category: AnimalCategory!
  }

  """
  Pagination metadata returned with any paginated list.
  Indicates whether more data exists and provides cursors for navigation.
  """
  type Pagination {
    hasNext: Boolean!
    hasPrevious: Boolean!
    startCursor: Cursor
    endCursor: Cursor
    """
    Optional total count for the dataset matching the current query/filter.
    May be expensive to compute; implement only if performance allows.
    """
    totalCount: Int
  }

  """
  Common cursor-based pagination arguments.
  Use either forward pagination (first/after) or backward pagination (last/before), not both.
  """
  input PageArgs {
    first: Int
    after: Cursor
    last: Int
    before: Cursor
  }

  """
  Sort direction for ordered results.
  """
  enum SortDirection {
    ASC
    DESC
  }

  """
  Sortable fields for countries.
  Used with CountryOrder to control result ordering.
  """
  enum CountryOrderField {
    NAME
    POPULATION
    AREA
    ID
  }

  """
  Defines ordering of country results.
  """
  input CountryOrder {
    field: CountryOrderField! = NAME
    direction: SortDirection! = ASC
  }

  """
  Sortable fields for animals.
  Used with AnimalOrder to control result ordering.
  """
  enum AnimalOrderField {
    NAME
    SPECIES
    CATEGORY
    ID
  }

  """
  Defines ordering of animal results.
  """
  input AnimalOrder {
    field: AnimalOrderField! = NAME
    direction: SortDirection! = ASC
  }

  """
  Filters for narrowing country search results.
  All fields are optional; only supplied filters will be applied.
  """
  input CountryFilter {
    continent: Continent
    populationMin: Int
    populationMax: Int
    areaMin: Float
    areaMax: Float
    name: String
    capital: String
    currency: String
  }

  """
  Filters for narrowing animal search results.
  All fields are optional; only supplied filters will be applied.
  """
  input AnimalFilter {
    category: AnimalCategory
    species: String
    habitat: String
    diet: String
    conservation_status: String
    name: String
  }

  """
  Paginated list of countries, including the data and pagination metadata.
  """
  type CountryPage {
    data: [Country!]!
    pagination: Pagination!
  }

  """
  Paginated list of animals, including the data and pagination metadata.
  """
  type AnimalPage {
    data: [Animal!]!
    pagination: Pagination!
  }

  type Query {
    """
    Returns a paginated list of countries with optional search, filtering, and ordering.
    """
    countriesPaginated(
      search: String,
      filter: CountryFilter,
      orderBy: CountryOrder = { field: NAME, direction: ASC },
      args: PageArgs = { first: 20 }
    ): CountryPage!

    """
    Returns a paginated list of animals with optional search, filtering, and ordering.
    """
    animalsPaginated(
      search: String,
      filter: AnimalFilter,
      orderBy: AnimalOrder = { field: NAME, direction: ASC },
      args: PageArgs = { first: 20 }
    ): AnimalPage!

    """
    Returns all countries matching optional search and filters.
    Deprecated in favor of countriesPaginated for performance and scalability.
    """
    countries(search: String, filter: CountryFilter): [Country!]! @deprecated(reason: "Use countriesPaginated")

    """
    Returns a single country by its unique ID.
    """
    country(id: ID!): Country

    """
    Returns all animals matching optional search and filters.
    Deprecated in favor of animalsPaginated for performance and scalability.
    """
    animals(search: String, filter: AnimalFilter): [Animal!]! @deprecated(reason: "Use animalsPaginated")

    """
    Returns a single animal by its unique ID.
    """
    animal(id: ID!): Animal
  }

  type Mutation {
    """
    Creates a new country with the specified details.
    """
    createCountry(
      name: String!,
      capital: String,
      population: Int,
      area: Float,
      currency: String,
      continent: Continent!
    ): Country!

    """
    Updates an existing country by its ID with provided fields.
    """
    updateCountry(
      id: ID!,
      name: String,
      capital: String,
      population: Int,
      area: Float,
      currency: String,
      continent: Continent
    ): Country

    """
    Deletes a country by its unique ID.
    """
    deleteCountry(id: ID!): Country

    """
    Creates a new animal with the specified details.
    """
    createAnimal(
      name: String!,
      species: String,
      habitat: String,
      diet: String,
      conservation_status: String,
      category: AnimalCategory!
    ): Animal!

    """
    Updates an existing animal by its ID with provided fields.
    """
    updateAnimal(
      id: ID!,
      name: String,
      species: String,
      habitat: String,
      diet: String,
      conservation_status: String,
      category: AnimalCategory
    ): Animal

    """
    Deletes an animal by its unique ID.
    """
    deleteAnimal(id: ID!): Animal
  }
`;
