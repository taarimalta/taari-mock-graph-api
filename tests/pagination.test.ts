import { ApolloServer } from 'apollo-server';
import { typeDefs } from '../src/schema/typeDefs';
import { countryResolvers } from '../src/resolvers/country';
import { animalResolvers } from '../src/resolvers/animal';

const server = new ApolloServer({
  typeDefs,
  resolvers: [countryResolvers, animalResolvers],
});

describe('Pagination API', () => {
  it('should paginate countries (first page)', async () => {
    const res = await server.executeOperation({
      query: `query { countriesPaginated(args: { first: 2 }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(res.data?.countriesPaginated.data.length).toBeLessThanOrEqual(2);
    expect(res.data?.countriesPaginated.pagination).toHaveProperty('hasNext');
    expect(res.data?.countriesPaginated.pagination).toHaveProperty('startCursor');
    expect(res.data?.countriesPaginated).toMatchSnapshot();
  });

  it('should paginate animals (first page)', async () => {
    const res = await server.executeOperation({
      query: `query { animalsPaginated(args: { first: 2 }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(res.data?.animalsPaginated.data.length).toBeLessThanOrEqual(2);
    expect(res.data?.animalsPaginated.pagination).toHaveProperty('hasNext');
    expect(res.data?.animalsPaginated.pagination).toHaveProperty('startCursor');
    expect(res.data?.animalsPaginated).toMatchSnapshot();
  });

  it('should return empty array for empty result', async () => {
    const res = await server.executeOperation({
      query: `query { countriesPaginated(filter: { name: "ZZZZZZ" }, args: { first: 2 }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(res.data?.countriesPaginated.data).toEqual([]);
    expect(res.data?.countriesPaginated).toMatchSnapshot();
  });

  it('should navigate forward using after cursor for countries', async () => {
    // Get first page and extract endCursor
    const firstRes = await server.executeOperation({
      query: `query { countriesPaginated(args: { first: 2 }) { data { id name } pagination { endCursor } } }`,
    });
    const afterCursor = firstRes.data?.countriesPaginated.pagination.endCursor;
    expect(afterCursor).toBeTruthy();

    // Get next page using after cursor
    const nextRes = await server.executeOperation({
      query: `query { countriesPaginated(args: { first: 2, after: "${afterCursor}" }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(nextRes.data?.countriesPaginated.data.length).toBeLessThanOrEqual(2);
    expect(nextRes.data?.countriesPaginated.pagination.hasPrevious).toBe(true);
    expect(nextRes.data?.countriesPaginated).toMatchSnapshot();
  });

  it('should navigate backward using before cursor for countries', async () => {
    // Get first page and extract endCursor
    const firstRes = await server.executeOperation({
      query: `query { countriesPaginated(args: { first: 2 }) { data { id name } pagination { endCursor } } }`,
    });
    const afterCursor = firstRes.data?.countriesPaginated.pagination.endCursor;
    expect(afterCursor).toBeTruthy();

    // Get next page using after cursor
    const nextRes = await server.executeOperation({
      query: `query { countriesPaginated(args: { first: 2, after: "${afterCursor}" }) { data { id name } pagination { endCursor } } }`,
    });
    const beforeCursor = nextRes.data?.countriesPaginated.pagination.endCursor;
    expect(beforeCursor).toBeTruthy();

    // Get previous page using before cursor (backward pagination)
    const prevRes = await server.executeOperation({
      query: `query { countriesPaginated(args: { last: 2, before: "${beforeCursor}" }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(prevRes.data?.countriesPaginated.data.length).toBeLessThanOrEqual(2);
    expect(prevRes.data?.countriesPaginated.pagination.hasNext).toBe(true);
    expect(prevRes.data?.countriesPaginated).toMatchSnapshot();
  });

  it('should navigate forward using after cursor for animals', async () => {
    const firstRes = await server.executeOperation({
      query: `query { animalsPaginated(args: { first: 2 }) { data { id name } pagination { endCursor } } }`,
    });
    const afterCursor = firstRes.data?.animalsPaginated.pagination.endCursor;
    expect(afterCursor).toBeTruthy();

    const nextRes = await server.executeOperation({
      query: `query { animalsPaginated(args: { first: 2, after: "${afterCursor}" }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(nextRes.data?.animalsPaginated.data.length).toBeLessThanOrEqual(2);
    expect(nextRes.data?.animalsPaginated.pagination.hasPrevious).toBe(true);
    expect(nextRes.data?.animalsPaginated).toMatchSnapshot();
  });

  it('should navigate backward using before cursor for animals', async () => {
    const firstRes = await server.executeOperation({
      query: `query { animalsPaginated(args: { first: 2 }) { data { id name } pagination { endCursor } } }`,
    });
    const afterCursor = firstRes.data?.animalsPaginated.pagination.endCursor;
    expect(afterCursor).toBeTruthy();

    const nextRes = await server.executeOperation({
      query: `query { animalsPaginated(args: { first: 2, after: "${afterCursor}" }) { data { id name } pagination { endCursor } } }`,
    });
    const beforeCursor = nextRes.data?.animalsPaginated.pagination.endCursor;
    expect(beforeCursor).toBeTruthy();

    const prevRes = await server.executeOperation({
      query: `query { animalsPaginated(args: { last: 2, before: "${beforeCursor}" }) { data { id name } pagination { hasNext hasPrevious startCursor endCursor totalCount } } }`,
    });
    expect(prevRes.data?.animalsPaginated.data.length).toBeLessThanOrEqual(2);
    expect(prevRes.data?.animalsPaginated.pagination.hasNext).toBe(true);
    expect(prevRes.data?.animalsPaginated).toMatchSnapshot();
  });
});
