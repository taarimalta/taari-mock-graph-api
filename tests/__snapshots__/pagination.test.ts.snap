// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Pagination API should navigate backward using before cursor for animals 1`] = `
{
  "data": [
    {
      "id": "4",
      "name": "Cheetah",
    },
    {
      "id": "26",
      "name": "Catfish",
    },
  ],
  "pagination": {
    "endCursor": "eyJpZCI6MjYsIm9yZGVyRmllbGQiOiJuYW1lIiwib3JkZXJWYWx1ZSI6IkNhdGZpc2giLCJkaXJlY3Rpb24iOiJERVNDIn0=",
    "hasNext": true,
    "hasPrevious": true,
    "startCursor": "eyJpZCI6NCwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiQ2hlZXRhaCIsImRpcmVjdGlvbiI6IkRFU0MifQ==",
    "totalCount": 30,
  },
}
`;

exports[`Pagination API should navigate backward using before cursor for countries 1`] = `
{
  "data": [
    {
      "id": "4",
      "name": "United States",
    },
    {
      "id": "1",
      "name": "Nigeria",
    },
  ],
  "pagination": {
    "endCursor": "eyJpZCI6MSwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiTmlnZXJpYSIsImRpcmVjdGlvbiI6IkRFU0MifQ==",
    "hasNext": true,
    "hasPrevious": false,
    "startCursor": "eyJpZCI6NCwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiVW5pdGVkIFN0YXRlcyIsImRpcmVjdGlvbiI6IkRFU0MifQ==",
    "totalCount": 6,
  },
}
`;

exports[`Pagination API should navigate forward using after cursor for animals 1`] = `
{
  "data": [
    {
      "id": "9",
      "name": "Bat",
    },
    {
      "id": "3",
      "name": "Bobcat",
    },
  ],
  "pagination": {
    "endCursor": "eyJpZCI6Mywib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiQm9iY2F0IiwiZGlyZWN0aW9uIjoiQVNDIn0=",
    "hasNext": true,
    "hasPrevious": true,
    "startCursor": "eyJpZCI6OSwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiQmF0IiwiZGlyZWN0aW9uIjoiQVNDIn0=",
    "totalCount": 30,
  },
}
`;

exports[`Pagination API should navigate forward using after cursor for countries 1`] = `
{
  "data": [
    {
      "id": "2",
      "name": "China",
    },
    {
      "id": "3",
      "name": "France",
    },
  ],
  "pagination": {
    "endCursor": "eyJpZCI6Mywib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiRnJhbmNlIiwiZGlyZWN0aW9uIjoiQVNDIn0=",
    "hasNext": true,
    "hasPrevious": true,
    "startCursor": "eyJpZCI6Miwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiQ2hpbmEiLCJkaXJlY3Rpb24iOiJBU0MifQ==",
    "totalCount": 6,
  },
}
`;

exports[`Pagination API should paginate animals (first page) 1`] = `
{
  "data": [
    {
      "id": "19",
      "name": "Axolotl",
    },
    {
      "id": "10",
      "name": "Bald Eagle",
    },
  ],
  "pagination": {
    "endCursor": "eyJpZCI6MTAsIm9yZGVyRmllbGQiOiJuYW1lIiwib3JkZXJWYWx1ZSI6IkJhbGQgRWFnbGUiLCJkaXJlY3Rpb24iOiJBU0MifQ==",
    "hasNext": true,
    "hasPrevious": false,
    "startCursor": "eyJpZCI6MTksIm9yZGVyRmllbGQiOiJuYW1lIiwib3JkZXJWYWx1ZSI6IkF4b2xvdGwiLCJkaXJlY3Rpb24iOiJBU0MifQ==",
    "totalCount": 30,
  },
}
`;

exports[`Pagination API should paginate countries (first page) 1`] = `
{
  "data": [
    {
      "id": "6",
      "name": "Australia",
    },
    {
      "id": "5",
      "name": "Brazil",
    },
  ],
  "pagination": {
    "endCursor": "eyJpZCI6NSwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiQnJhemlsIiwiZGlyZWN0aW9uIjoiQVNDIn0=",
    "hasNext": true,
    "hasPrevious": false,
    "startCursor": "eyJpZCI6Niwib3JkZXJGaWVsZCI6Im5hbWUiLCJvcmRlclZhbHVlIjoiQXVzdHJhbGlhIiwiZGlyZWN0aW9uIjoiQVNDIn0=",
    "totalCount": 6,
  },
}
`;

exports[`Pagination API should return empty array for empty result 1`] = `
{
  "data": [],
  "pagination": {
    "endCursor": null,
    "hasNext": false,
    "hasPrevious": false,
    "startCursor": null,
    "totalCount": 0,
  },
}
`;
